Solving #186 Positive reinforcement

shaman-rajan opened the pull request:
I'm still working on editing the data schema to all questions, but created the PR to take comments in the meantime. 2 autosave karma tests fail, and I'm working on fixing that too.

shaman-rajan commented:
All tests pass now and all questions have tags included. @eyurko PTAL.

seanlip commented on assets/tests/TaskSchemaValidationService.js:
```diff
@@ -157,6 +157,12 @@ tie.factory('TaskSchemaValidationService', [
+      verifyCorrectnessTestsHaveTags: function(task) {
+        var correctnessTests = task.getCorrectnessTests();
+        return correctnessTests.every(function(test) {
+          return test.getTag() !== undefined && test.getTag().length > 0;
```
angular.isString(test.getTag()) is probably a stronger check.

seanlip commented on client/components/LearnerViewDirective.js:
```diff
@@ -481,6 +505,19 @@ tie.directive('learnerView', [function() {
+        QuestionDataService.initCurrentQuestionSet(questionSetId);
+        var questionSet = QuestionDataService.getCurrentQuestionSet(
+          questionSetId);
+        $scope.currentQuestionIndex = 0;
```
Don't copy/paste please -- refactor common stuff into a function instead. That said, I just made a similar comment on #250 which has just been merged, so merging from master into this branch and resolving the conflicts should address this.

shaman-rajan replied:
This isn't needed, will remove it.

seanlip commented on client/domain/ReinforcementBulletObjectFactory.js:
```diff
@@ -0,0 +1,57 @@
+    var ReinforcementBullet = function(passed, content) {
+      this._passed = passed;
+      this._content = content;
+      if (passed) {
```
this._imgName = (passed ? 'pass.png' : 'fail.png');

but then, actually, you don't even need this._imgName at all. Just add a method ReinforcementBullet.prototype.getImageFilename() and do the computation there.

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,98 @@
+  function() {
+
+    return {
+      getReinforcement: function(task, codeEvalResult, runtimeFeedback) {
```
It's an antipattern to add new keys to an object mid-stream -- let's avoid that. It means that the same-looking object has a different definition depending on which file in the code you're looking at.

Instead, construct the Reinforcement object here, and return just that.

seanlip commented on client/services/SolutionHandlerService.js:
```diff
@@ -60,7 +60,9 @@ tie.factory('SolutionHandlerService', [
             TranscriptService.recordSnapshot(
               SnapshotObjectFactory.create(codeEvalResult, runtimeFeedback));
-            return runtimeFeedback;
+            var finalFeedback = ReinforcementGeneratorService.getReinforcement(
```
Don't modify the feedback object. Instead, when passing stuff to the learner view directive, pass the reinforcement object should be passed as a separate object from the usual feedback. Do all the computations in services so that there is no logic in the directive controller.

shaman-rajan commented:
@seanlip PTAL. Comments addressed except the one we can fix during merge.

shaman-rajan commented:
Looks like there are still issues here. When submitting with a syntax error, the app hangs. Will update once it's fixed.

seanlip commented:
OK, thanks for the update @shaman-rajan! I'll take another look when it's ready.

shaman-rajan commented:
@seanlip PTAL. Fixed the syntax error issue, and addressed all comments.

seanlip commented on client/components/LearnerViewDirective.js:
```diff
@@ -318,6 +324,21 @@ tie.directive('learnerView', [function() {
+          margin-top: 1px;
+          position: relative;
+        }
+        .bullet-img {
```
Could you use `.tie-bullet-img` or `.tie-reinforcement .bullet-img` instead? We prefix all our CSS classes with `.tie-` in order to distinguish them from third-party libs.

seanlip commented on client/components/LearnerViewDirective.js:
```diff
@@ -318,6 +324,21 @@ tie.directive('learnerView', [function() {
+          position: relative;
+        }
+        .bullet-img {
+          width: 15px;
```
Nit: recommend alphabetizing these attrs (i.e. bottom, then height, then position, then width). As the number of attributes increases, this makes it easier to locate a particular one.

seanlip commented on client/components/LearnerViewDirective.js:
```diff
@@ -553,9 +588,28 @@ tie.directive('learnerView', [function() {
+              // Updating reinforcement bullets only if no syntax errors
+              reinforcement.clear();
+              for (var caseList in reinforcementObject.passedList) {
+                if (reinforcementObject.passedList[caseList]) {
```
The reinforcement object should be fully prepared in a service and just passed to the controller -- there should be none of this logic in the directive's controller.

If necessary, you can handle the "don't show if there are syntax errors" bit with an if statement -- but that's about as much logic as should exist at this level of the code.

shaman-rajan replied:
I was using a style similar to congratulatoryFeedback defined on line 515. The reinforcement object we get from the service is always a new one, from which entries are taken and the main stored object is updated.

seanlip replied:
Ah, I see. Let's not do that -- congratulatoryFeedback is a special case, and is just meant to show a message at the end of the task (which is just a display thing, and no logic is required). However, for something complex like reinforcement data, it's better to just pass the object wholesale from the service and do the logic within the service.

seanlip commented on client/domain/ReinforcementBulletObjectFactory.js:
```diff
@@ -0,0 +1,52 @@
+    };
+
+    ReinforcementBullet.prototype.getImgName = function() {
+      return this._passed ? "pass.png" : "fail.png";
```
Use single quotes here for consistency with the other files.

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,99 @@
+          var splitTests = {};
+          for (var i = 0; i < tests.length; ++i) {
+            var test = tests[i];
+            test.index = i;
```
Please avoid modifying dicts like this. If you change the keys midstream then it's hard to tell what a dict's structure is at any given time.

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,99 @@
+
+        // Initializing question reinforcement data if not done already
+        if (!task.passedList) {
+          task.passedList = {};
```
Please avoid modifying dicts like this. If you change the keys midstream then it's hard to tell what a dict's structure is at any given time.

Also, why is this defined on the task? I thought it would be part of the reinforcement object, conceptually?

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,99 @@
+            var test = tests[i];
+            test.index = i;
+            var testTag = test.getTag();
+            if (testTag in splitTests) {
```
Use splitTests.hasOwnProperty(testTag) for checking object membership.

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        var splitTestsByTag = function(tests) {
+          var splitTests = {};
```
testsByTag or testsKeyedByTag might be a better name.

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,99 @@
+          return splitTests;
+        };
+
+        // Go through correctness tests to update reinforcement data
```
Add a period at the end of the comment.

seanlip commented on client/services/SolutionHandlerService.js:
```diff
@@ -39,7 +39,10 @@ tie.factory('SolutionHandlerService', [
-            return $q.resolve(feedback);
+            return {
+              feedbackObject: feedback,
+              reinforcementObject: {}
```
This should be a blank Reinforcement object. Avoid passing dicts around.

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,99 @@
+        if (!task.passedList) {
+          task.passedList = {};
+        }
+        if (!task.pastFailsList) {
```
I think I understand why you're doing it like this -- you want to maintain history. However, defining stuff on the task dict midstream is not the right way to do this, since the task is supposed to be independent of the individual student's playthrough.

So, there are two ways we can go here:
(a) The "proper" way to do this is probably to preserve info about "past fails" somewhere. Perhaps you can store these in the transcript, and extract them from there.
(b) If you want to get this PR in quicker, though, that may be too much to do. In that case, let's drop the "past fails" requirement for this PR, and handle things in a stateless manner.

I'm happy either way. Let me know if you want to discuss any of this.

BenjaminHolland replied:
I kind of like the idea of having a more robust transcript system, especially for repeat playthroughs. Eventually we might be able to turn that into a progress or improvement system of sorts.

shaman-rajan replied:
I like the idea of adding it to the transcript. Will work on doing that.

seanlip commented:
Also, FYI, there are merge conflicts.

shaman-rajan commented:
@seanlip PTAL

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -0,0 +1,85 @@
+      return this._task;
+    };
+
+    Reinforcement.prototype.addToPassedList = function(key, value) {
```
What exactly do `key` and `value` represent here? (E.g. is the "key" a "tag"?)

Please add comments to explain, or use clearer variable names. Ditto below.

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -0,0 +1,85 @@
+      }
+
+      for (var testCase in this._pastFailsList) {
+        if (this._pastFailsList[testCase]) {
```
In principle, is it possible for testCase here to overlap a caseList in this._passedList? If so, how do you guard against that (within this object)?

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,101 @@
+    return {
+      getReinforcement: function(task, codeEvalResult) {
+
+        var testsKeyedByTag = function(tests) {
```
Move this between lines 22 and 23 -- no need for inner functions.

Also: getTestsKeyedByTag. Names of functions should start with verbs.

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,101 @@
+          var splitTests = {};
+          for (var i = 0; i < tests.length; ++i) {
+            var test = tests[i];
+            test.originalIndex = i;
```
Please do not modify objects midstream -- we talked about this before.

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,101 @@
+        // Copy the previous reinforcement object if we are on the same task.
+        if (previousReinforcement !== null &&
+            task === previousReinforcement.getTask()) {
+          for (var tag in previousReinforcement.getPassedList()) {
```
Define a local variable for previousReinforcement.getPassedList() and use that instead; don't keep calling the method repeatedly (especially in a for loop). Ditto for the past fails list.

It may make sense to add a reinforcement.clonePassedList() method that uses angular.copy() to bulk-populate the list; if you go that route, make sure to document the expected inputs.

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,101 @@
+        }
+
+        // Go through correctness tests to update reinforcement data.
+        var correctnessTests = testsKeyedByTag(task.getCorrectnessTests());
```
correctnessTestsByTag

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,101 @@
+            var test = correctnessTests[testTag][testIdx];
+            var observedOutput = observedOutputs[test.originalIndex];
+            if (test.matchesOutput(observedOutput)) {
+              if (test.getInput() in reinforcement.getPastFailsList()) {
```
reinforcement.hasPastFailure(test.getInput())

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,101 @@
+            var observedOutput = observedOutputs[test.originalIndex];
+            if (test.matchesOutput(observedOutput)) {
+              if (test.getInput() in reinforcement.getPastFailsList()) {
+                reinforcement.addToPastFailsList(test.getInput(), true);
```
But it's already in the list. Should this be "modify" instead?

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,101 @@
+                reinforcement.addToPastFailsList(test.getInput(), false);
+                failedCaseSeenOverall = true;
+              }
+              failedCaseSeenInTag = true;
```
I'm confused. If we change failedCaseSeenInTag to true when we see it is false, then what's the point at all of keeping track of failedCaseSeenInTag?

shaman-rajan replied:
This is inside an "else if" clause, so is entered only if the first if condition evaluates to false.

seanlip replied:
Ah. Perhaps, then:

    } else {
      failedCaseSeenInTag = true;
      if (!failedCaseSeenOverall) {
        ...
      }
    }

might be slightly less efficient, but clearer?

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,101 @@
+          }
+          if (failedCaseSeenInTag) {
+            if (testTag in reinforcement.getPassedList()) {
+              reinforcement.addToPassedList(testTag, false);
```
I'm starting to think that "passed list" is a bit of a misnomer here. What are your two lists tracking, exactly? Can the names better reflect their contents?

seanlip commented on client/services/TranscriptService.js:
```diff
@@ -27,6 +27,9 @@ tie.factory('TranscriptService', [
       recordSnapshot: function(snapshot) {
         transcript.recordSnapshot(snapshot);
+      },
+      getPreviousSnapshot: function() {
```
getMostRecentSnapshot() perhaps? "Previous" implies any previous one.

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -21,15 +21,15 @@ tie.factory('ReinforcementObjectFactory', [
       this._task = task;
-      this._passedList = {};
-      this._pastFailsList = {};
+      this._passedTagsList = {};
```
One other thing I forgot to mention earlier (sorry) -- can we call these this._passedTags and this._pastFailedCases? They're not lists; lists are ordered.

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -59,20 +59,24 @@ tie.factory('ReinforcementObjectFactory', [
+    };
+
+    Reinforcement.prototype.hasPastFailedCase = function(caseInput) {
+       return (caseInput in this._pastFailedCasesList);
```
return this._pastFailedCasesList.hasOwnProperty(caseInput)

Btw, are the caseInput's guaranteed to be strings / hashable?

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -59,20 +59,24 @@ tie.factory('ReinforcementObjectFactory', [
 
-    Reinforcement.prototype.addToPassedList = function(key, value) {
-      this._passedList[key] = value;
+    Reinforcement.prototype.addToPassedTagsList = function(key, value) {
```
Still need docstrings.

shaman-rajan commented:
@seanlip Addressed comments. 1 thing left to consider is whether all inputs are hashable. If we are using templates to enter test cases, they will initially be strings that may later be parsed to other data structures. I will investigate on whether the objects I am using to hash are strings taken from the template or parsed objects and update.

shaman-rajan commented:
@seanlip PTAL. Addressed all comments including the hashability of inputs.

seanlip commented on client/domain/CorrectnessTestObjectFactory.js:
```diff
@@ -24,13 +24,21 @@ tie.factory('CorrectnessTestObjectFactory', [
 
+    CorrectnessTest.prototype.getStringifiedInput = function() {
+      if (!this._stringifiedInput) {
+        this._stringifiedInput = JSON.stringify(this._input);
```
Again, do not modify dicts mid-stream. Either always compute on the fly, or store at the outset and do a simple lookup.

shaman-rajan replied:
I didn't want to store that unless this function is used. That said, I can do this computation during object creation also.

seanlip replied:
Let's either always do it at object creation time, or always compute it on the fly. I'm fine either way.

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -0,0 +1,104 @@
+      ReinforcementBulletObjectFactory) {
+    var Reinforcement = function(task) {
+      this._task = task;
+      this._passedTags = {};
```
Can you document the keys/values (and their types) for these two objects?

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -0,0 +1,104 @@
+    };
+
+    // Add a tag to the list of tags the code passes/fails
+    Reinforcement.prototype.addToPassedTags = function(tag, passedBool) {
```
addPassedTag

seanlip replied:
Also, here and below, would testPassed be a better name for the boolean? In general you'll notice that the variable/arg names don't include type information. It seems like a good idea to point out that this is a boolean but I would recommend doing it in the comment/docstring.

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -0,0 +1,104 @@
+    };
+
+    // Update the pass/fail status of a tag
+    Reinforcement.prototype.updatePassedTags = function(tag, passedBool) {
```
updatePassedTag

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -0,0 +1,104 @@
+
+    // Add a test case to the list of failed test cases
+    Reinforcement.prototype.addToPastFailedCases = function(
+        caseInput, passedBool) {
```
The comments here and below should indicate that caseInput is a string, or perhaps you can rename to stringifiedInput.

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -0,0 +1,104 @@
+    };
+
+    Reinforcement.prototype.hasPastFailedCase = function(caseInput) {
+      return (caseInput in this._pastFailedCases);
```
return this._pastFailedCases.hasOwnProperty(stringifiedInput);

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,120 @@
+
+    var getTestsKeyedByTagAndIndexes = function(tests) {
+      var splitTests = {};
+      var inputToOriginalIndex = {};
```
I think it's worth adding comments to explain what the keys/values of these two objects are, and why they're needed. The names are reasonable but a bit hard to parse for someone who doesn't have context.

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,120 @@
+          var previousReinforcementPassedTags =
+            previousReinforcement.getPassedTags();
+          for (var tag in previousReinforcementPassedTags) {
+            reinforcement.addToPassedTags(tag,
```
Here and below: if a line of code spans multiple lines, break after "(".

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,120 @@
+        }
+      }
+      return {
+        splitTests: splitTests,
```
Avoid returning composite blobs like this where possible. In this case, have two separate methods instead, each of which only returns one thing.

seanlip commented on client/services/ReinforcementGeneratorService.js:
```diff
@@ -0,0 +1,120 @@
+        var testIndexes = correctnessTestsByTagAndIndexes.testIndexes;
+        var observedOutputs = codeEvalResult.getCorrectnessTestResults();
+        // Get results only for the last task.
+        observedOutputs = observedOutputs[observedOutputs.length - 1];
```
Don't redefine a variable. Instead, create a new method along the lines of codeEvalResult.getLastTaskResults()

shaman-rajan commented:
@seanlip PTAL

seanlip commented on client/domain/CorrectnessTestObjectFactory.js:
```diff
@@ -22,15 +22,21 @@ tie.factory('CorrectnessTestObjectFactory', [
     };
 
+    CorrectnessTest.prototype.getStringifiedInput = function() {
+      return JSON.stringify(this._input);
```
If you are storing the stringified input, and if the input doesn't change, why are you redoing the serialization?

In other words, either do this, or keep line 25, but not both.

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -0,0 +1,120 @@
+    *                   reinforcement comments and the corresponding pass/fail
+    *                   status.
+    *                   Key: Stringified input of test case.
+    *                   Value: Boolean indicating where the code passed the
```
where --> whether?

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -0,0 +1,120 @@
+    * Fields:
+    *  task: The corresponding task for the reinforcement comments.
+    *  passedTags: Dictionary containing tags (test sets) that the code
+    *              passed/failed.
```
Add: A tag is considered "failed" if at least one of the correctness tests corresponding to the tag have failed.

seanlip commented on client/domain/ReinforcementObjectFactory.js:
```diff
@@ -0,0 +1,120 @@
+        }
+      }
+
+      for (var testCase in this._pastFailedCases) {
```
var stringifiedTestCase

Recommend similar changes elsewhere throughout this file (and possibly others), if necessary, so that it's always clear whether we're dealing with a stringified version or the raw version.

shaman-rajan commented:
@seanlip PTAL

seanlip commented:
@shaman-rajan -- this looks great to me, and I'm going to merge it. Thank you very much for all your work on this!
